writer.go

```diff
@@ -0,0 +1,246 @@
+	for i := 0; i < v.NumField(); i++ {
+		// Get the field tag value
+		rawTag := v.Type().Field(i).Tag.Get(tagName)
+		if rawTag == rsfIgnore {
+			continue
+		}
+
+		t := &tag{}
+		var skip bool
+
+		fieldVal := v.Field(i)
+		if rawTag != "" {
+			tagParts := strings.Split(rawTag, rsfDelim)
+			t.name = tagParts[0]
+			for j := 1; j < len(tagParts); j++ {
+				part := strings.TrimSpace(strings.ToLower(tagParts[j]))
+				if part == rsfSkip {
+					skip = true
+				}
+				if strings.HasPrefix(part, rsfIndex+rsfSep) && len(part) > 6 {
+					indexParts := strings.Split(part, rsfSep)
+					t.index = indexParts[1]
+				}
+				if strings.HasPrefix(part, rsfFixed+rsfSep) && len(part) > 6 {
+					fixedParts := strings.Split(part, rsfSep)
+					var err error
+					t.fixed, err = strconv.Atoi(fixedParts[1])
+					if err != nil {
+						return err
+					}
+				}
+			}
+			if tParent.index == t.name {
+				tParent.indexVal = fieldVal.String()
+				tParent.indexSz = t.fixed
+			}
+		}
+		if !skip {
+			err := f.writeObject(fieldVal, t, buf)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (f *writer) writeArray(v reflect.Value, t *tag, buf *bytes.Buffer) error {
+	var snapIndexBuf *bytes.Buffer
+	var snapBuf *bytes.Buffer
+	if t.index != "" {
+		snapIndexBuf = &bytes.Buffer{}
+		snapBuf = &bytes.Buffer{}
+	} else {
+		snapBuf = buf
+	}
+	_, err := f.WriteSizeField(0, v.Len(), buf)
+	if err != nil {
+		return err
+	}
+	var lastLen int
+	for i := 0; i < v.Len(); i++ {
+		el := v.Index(i)
+		err = f.writeObject(el, t, snapBuf)
+		if err != nil {
+			return err
+		}
+		bufLen := snapBuf.Len()
+		if t.index != "" {
+			_, err = f.WriteFixedStringField(0, t.indexSz, t.indexVal, snapIndexBuf)
+			if err != nil {
+				return err
+			}
+			_, err = f.WriteSizeField(0, bufLen-lastLen, snapIndexBuf)
+			if err != nil {
+				return err
+			}
+			lastLen = bufLen
+		}
+	}
+	if t.index != "" {
+		_, err = io.Copy(buf, snapIndexBuf)
+		if err != nil {
+			return err
+		}
+		_, err = io.Copy(buf, snapBuf)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (f *writer) writeString(s string, t *tag, buf *bytes.Buffer) error {
+	var err error
+	if t.fixed > 0 {
+		_, err = f.WriteFixedStringField(0, t.fixed, s, buf)
+	} else {
+		_, err = f.WriteStringField(0, s, buf)
+	}
+	return err
+}
+
+func (f *writer) WriteSizeField(pos int, val int, r io.Writer) (int, error) {
+	var i, sz int
+	var err error
+
+	// Write size
+	bs := make([]byte, 4)
+	binary.LittleEndian.PutUint32(bs, uint32(val))
+	i, err = r.Write(bs)
+	if err != nil {
+		return 0, err
+	}
+	sz += i
+
+	return pos + sz, nil
+}
+
+func (f *writer) WriteFixedStringField(pos, sz int, val string, r io.Writer) (int, error) {
+	var i int
+	var err error
+
+	if sz != len(val) {
+		return 0, fmt.Errorf("size %d does not match expected size %d", len(val), sz)
+	}
+
+	// Write value
+	i, err = r.Write([]byte(val))
```

Comment:
I don't think I follow why `r io.Writer` is passed in to these methods. Why aren't they using the `io.Writer` from the `writer` struct?

Reply:
I can't remember all the reasons, but it's actually worked out well in practice. One interesting thing I ran into just today is a good example:

In the code that reads packages (not in this library, currently), we use a `bufio.Reader` with a fixed buffer size to speed up reading considerably. However, a `bufio.reader` doesn't support `Seek`. In some cases, it's faster to call `Seek` on the actual `*os.File` object and then reset the associated `bufio.Reader`. If we tried to incapsulate the reader here, we wouldn't have control over things like this, and it would also make it more difficult to tweak things like the buffer size.

Reply:
Makes sense, thanks for explaining!